File: ta_lib/strategies/signal/src/ma/ma2_rsi.rs
use base::prelude::*;
use core::prelude::*;
use indicator::{ma_indicator, MovingAverageType};
use momentum::rsi;
use timeseries::prelude::*;

const RSI_UPPER_BARRIER: f32 = 85.0;
const RSI_LOWER_BARRIER: f32 = 15.0;

pub struct Ma2RsiSignal {
    source_type: SourceType,
    smooth_type: Smooth,
    rsi_period: usize,
    threshold: f32,
    ma: MovingAverageType,
    fast_period: usize,
    slow_period: usize,
}

impl Ma2RsiSignal {
    pub fn new(
        source_type: SourceType,
        smooth_type: Smooth,
        rsi_period: f32,
        threshold: f32,
        ma: MovingAverageType,
        fast_period: f32,
        slow_period: f32,
    ) -> Self {
        Self {
            source_type,
            smooth_type,
            rsi_period: rsi_period as usize,
            threshold,
            ma,
            fast_period: fast_period as usize,
            slow_period: slow_period as usize,
        }
    }
}

impl Signal for Ma2RsiSignal {
    fn lookback(&self) -> usize {
        let adj_lookback = std::cmp::max(self.fast_period, self.slow_period);
        std::cmp::max(adj_lookback, self.rsi_period)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let rsi = rsi(
            &data.source(self.source_type),
            self.smooth_type,
            self.rsi_period,
        );

        let ma_short = ma_indicator(&self.ma, data, self.source_type, self.fast_period);
        let ma_long = ma_indicator(&self.ma, data, self.source_type, self.slow_period);

        let lower_barrier = RSI_LOWER_BARRIER + self.threshold;
        let upper_barrier = RSI_UPPER_BARRIER - self.threshold;

        (
            data.close().sgt(&ma_short)
                & data.close().sgt(&ma_long)
                & ma_short.sgt(&ma_long)
                & rsi.cross_under(&lower_barrier),
            data.close().slt(&ma_short)
                & data.close().slt(&ma_long)
                & ma_short.slt(&ma_long)
                & rsi.cross_over(&upper_barrier),
        )
    }
}

File: factor/signal/ma/ma2_rsi.py
from dataclasses import dataclass

from core.models.moving_average import MovingAverageType
from core.models.parameter import (
    CategoricalParameter,
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from core.models.source import SourceType
from factor.signal.base import Signal, SignalType


@dataclass(frozen=True)
class Ma2RsiSignal(Signal):
    type: SignalType = SignalType.Ma2Rsi
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    smooth_type: Parameter = StaticParameter(Smooth.SMMA)
    rsi_period: Parameter = StaticParameter(2.0)
    threshold: Parameter = RandomParameter(0.0, 3.0, 1.0)
    ma: Parameter = CategoricalParameter(MovingAverageType)
    fast_period: Parameter = RandomParameter(20.0, 50.0, 5.0)
    slow_period: Parameter = RandomParameter(30.0, 50.0, 5.0)
